assignments/walk_tests_repo.go

```diff
@@ -0,0 +1,156 @@
+package assignments
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sort"
+
+	pb "github.com/autograde/quickfeed/ag"
+)
+
+const (
+	assignmentFile     = "assignment.yml"
+	assignmentFileYaml = "assignment.yaml"
+	criteriaFile       = "criteria.json"
+	scriptTemplateFile = "run.sh"
+	scriptFolder       = "scripts"
+	dockerfile         = "Dockerfile"
+	taskFilePattern    = "task-*.md"
+)
+
+var patterns = []string{
+	assignmentFile,
+	assignmentFileYaml,
+	criteriaFile,
+	scriptTemplateFile,
+	dockerfile,
+	taskFilePattern,
+}
+
+// matchAny returns true if filename matches one of the target patterns.
+func matchAny(filename string) bool {
+	for _, pattern := range patterns {
+		if ok, _ := filepath.Match(pattern, filename); ok {
+			return true
+		}
+	}
+	return false
+}
+
+// match returns true if filename matches the given pattern.
+func match(filename, pattern string) bool {
+	if ok, _ := filepath.Match(pattern, filename); ok {
+		return true
+	}
+	return false
+}
+
+// readTestsRepositoryContent reads dir and returns a list of assignments and the course's Dockerfile.
+func readTestsRepositoryContent(dir string, courseID uint64) ([]*pb.Assignment, string, error) {
+	files, err := walkTestsRepository(dir)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Process all assignment.yml files first
+	assignmentsMap := make(map[string]*pb.Assignment)
+	for path, contents := range files {
+		assignmentName := filepath.Base(filepath.Dir(path))
+		switch filepath.Base(path) {
+		case assignmentFile, assignmentFileYaml:
+			assignment, err := newAssignmentFromFile(contents, assignmentName, courseID)
+			if err != nil {
+				return nil, "", err
+			}
+			assignmentsMap[assignmentName] = assignment
+		}
+	}
+
+	var defaultScriptTemplate string
+	var courseDockerfile string
+
+	// Process other files in tests repository
+	for path, contents := range files {
+		assignmentName := filepath.Base(filepath.Dir(path))
+
+		switch filepath.Base(path) {
+		case criteriaFile:
+			var benchmarks []*pb.GradingBenchmark
+			if err := json.Unmarshal(contents, &benchmarks); err != nil {
+				return nil, "", fmt.Errorf("failed to unmarshal %q: %s", criteriaFile, err)
+			}
+			assignmentsMap[assignmentName].GradingBenchmarks = benchmarks
+
+		case scriptTemplateFile:
+			if assignmentName != scriptFolder {
+				// Found assignment-specific script template
+				assignmentsMap[assignmentName].ScriptFile = string(contents)
+			} else {
+				defaultScriptTemplate = string(contents)
+			}
+
+		case dockerfile:
+			courseDockerfile = string(contents)
+		}
+
+		if match(filepath.Base(path), taskFilePattern) {
+			assignment := assignmentsMap[assignmentName]
+			// Probably a better way of acquiring task name
+			task, err := newTask(contents, assignment, assignmentName+"/"+filepath.Base(path))
```

Comment:
Assuming you add the `taskName()` helper function in `tasks.go` you can do this:
```go
	taskName := taskName(assignmentName, filepath.Base(path))
	task, err := newTask(contents, assignment, taskName)
```
Or better avoid passing `assignment` to `newTask` by passing instead the `uint32`: `assignment.GetOrder()`. See `tasks.go` for additional comments.